### PR TITLE
Support interface revert a rejection

### PR DESCRIPTION
**Pull request by stevehook**
## Context

This PR will allow support users to revert accidental rejections by a provider. To revert a rejection a support user must enter a support ticket URL.

The design used in this PR is based on https://github.com/DFE-Digital/apply-for-teacher-training/pull/4572.

## Changes proposed in this pull request

- [x] Add a new _Revert rejection_ link to the status line on application page in the support interface.
image

- [x] The above link is to a new confirmation page:
image

- [x] Implement a form and service backing the UI above.
- [x] New system spec

## Guidance to review

- The content for the confirmation is adapted from the reinstate offer feature. It needs some work, in particular I don't know what the guidance should be notifications for candidate and provider.

## Link to Trello card

https://trello.com/c/LPKMNuUN/3361-add-functionality-to-the-support-ui-to-revert-a-rejection

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/system/support_interface/revert_rejection_spec.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+require 'rails_helper'
+
+RSpec.feature 'Revert an accidental rejection' do
+  include DfESignInHelpers
+
+  scenario 'View structured reasons for rejection', with_audited: true do
```
Comment (malcolmbaig) [inline spec/system/support_interface/revert_rejection_spec.rb]: nit: worth correcting the scenario here so the commit makes sense in isolation (can see it's corrected later on)

File: spec/components/support_interface/application_choice_component_spec.rb [hunk 2]
```diff
@@ -39,6 +39,32 @@
...
+      expect(result.css('.govuk-summary-list__actions a')[0].attr('href')).to include(
+        Rails.application.routes.url_helpers.support_interface_application_form_revert_rejection_path(
+          application_form_id: rejected_application_choice.application_form.id,
+          application_choice_id: rejected_application_choice.id,
+        ),
+      )
+      expect(result.css('.govuk-summary-list__actions').text.strip).to include('Revert rejection')
+    end
+
+    it 'Does not renders a link to the revert rejection page when application was rejected by default' do
```
Comment (malcolmbaig) [inline spec/components/support_interface/application_choice_component_spec.rb:57]: typo: renders

File: spec/services/support_interface/revert_rejection_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      described_class.new(
+        application_choice: application_choice,
+        zendesk_ticket: zendesk_ticket,
+      ).save!
+
+      expect(application_choice).to be_awaiting_provider_decision
+      expect(application_choice.audits.last.comment).to include(zendesk_ticket)
+    end
+  end
+end
```
Comment (malcolmbaig) [inline spec/services/support_interface/revert_rejection_spec.rb:18]: Quite liked the testing pattern used [here](https://github.com/DFE-Digital/apply-for-teacher-training/pull/4572/files#diff-6e3f92ab4664c6b3c021cb2511b292d4a484c9c4acec2bde45f3ba7f1dfda083) where DeclineOffer is used to make the state change and we expect equality between the original choice and the updated one. Would it make sense to do similar here?

**Review by malcolmbaig — APPROVED** [review]
Code looks good, and works as expected in review app. A few minor comments left inline. Maybe needs a designer to take a look as well.

File: app/views/support_interface/application_forms/application_choices/confirm_revert_rejection.html.erb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <%= f.govuk_error_summary %>
+      <h1 class="govuk-heading-l">
+        Are you sure you want to revert this rejection?
+      </h1>
+
+      <p class="govuk-body">A rejection can only be reverted if:</p>
+      <ul class="govuk-list govuk-list--bullet">
+        <li>it was rejected within 5 working days of the request to revert and,</li>
```
Comment (paulrobertlloyd) [inline app/views/support_interface/application_forms/application_choices/confirm_revert_rejection.html.erb:18]:
```suggestion
        <li>it was rejected within 5 working days of the request to revert</li>
```

File: app/views/support_interface/application_forms/application_choices/confirm_revert_rejection.html.erb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      <p class="govuk-body">Once the rejection has been reverted, please email the candidate to let them know.</p>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <%= f.govuk_text_field(
+              :audit_comment_ticket,
+              label: {
+                text: 'Zendesk ticket URL',
```
Comment (paulrobertlloyd) [inline app/views/support_interface/application_forms/application_choices/confirm_revert_rejection.html.erb:31]: Given we are using the same label (and hint) value for auditing changes like this, should we localise these strings to me this a bit DRYer?

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]
Just a few small things.

**Comment by stevehook** [comment]
> Oh, I was thinking all occurrences use one set of strings? Is that a good idea?

I wasn't sure about this. There is a school of thought that says that duplication in i18n files is ok because different things in different contexts may need different content. But I don't think that applies here so I've moved them all to more a generic key.

**Review by paulrobertlloyd — APPROVED** [review]

**Pull request merged**